tests/apollo/test_skvbc_checkpoints.py

```diff
@@ -123,6 +123,106 @@ async def test_checkpoint_propagation_after_crashed_replicas_comeback(self, bft_
+        6) Verify view change has occurred
+        7) Verify checkpoint propagation to current primary in the new view
+        8) Bring the f replicas up
+        9) Verify checkpoint propagation to a stale node of the f replicas brought up
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        n = bft_network.config.n
+        f = bft_network.config.f
+        c = bft_network.config.c
+
+        self.assertEqual(len(bft_network.procs), n,
+                         "Make sure all replicas are up initially.")
+
+        initial_primary = await bft_network.get_current_primary()
+
+        expected_next_primary = initial_primary + 1
+
+        checkpoint_init_primary_before = await bft_network.wait_for_checkpoint(replica_id=initial_primary)
+
+        # crashing f-1 replicas, initial_primary is crashed after checkpoint creation to trigger vc
+        crashed_replicas = await self._crash_replicas(
+            bft_network=bft_network,
+            nb_crashing=f - 1,
+            exclude_replicas={expected_next_primary, initial_primary}
+        )
+        self.assertFalse({initial_primary, expected_next_primary}.issubset(crashed_replicas))
+
+        self.assertGreaterEqual(
+            len(bft_network.procs), 2 * f + 2 * c + 1,
+            "Make sure enough replicas are up to allow a successful view change!")
+
+        await skvbc.fill_and_wait_for_checkpoint(
+            initial_nodes=bft_network.all_replicas(without=crashed_replicas),
+            checkpoint_num=1,
+            verify_checkpoint_persistency=False
+        )
+
+        # verify checkpoint creation at the initial primary after f - 1 replicas down
+        checkpoint_init_primary_after = await bft_network.wait_for_checkpoint(replica_id=initial_primary)
+        self.assertEqual(checkpoint_init_primary_after, 1 + checkpoint_init_primary_before)
+
+        await bft_network.wait_for_view(
+            replica_id=initial_primary,
+            expected=lambda v: v == initial_primary,
+            err_msg="Make sure we are in the initial view "
+                    "before crashing the primary."
+        )
+
+        # trigger a view change by crashing the initial primary
+        bft_network.stop_replica(initial_primary)
+        crashed_replicas.add(initial_primary)
+
+        try:
+            with trio.move_on_after(seconds=1):  # seconds
+                await skvbc.send_indefinite_write_requests()
+
+        except trio.TooSlowError:
+            pass
+
+        finally:
+            # wait for view change
+            await bft_network.wait_for_view(
+                replica_id=random.choice(bft_network.all_replicas(without=crashed_replicas)),
+                expected=lambda v: v == expected_next_primary,
+                err_msg="Make sure view change has been triggered."
+            )
+
+        current_primary = initial_primary + 1
+        self.assertEqual(current_primary, expected_next_primary)
```

Comment:
Sorry, I confused you, it should rather be:
`self.assertFalse(initial_primary in crashed_replicas or expected_next_primary in crashed_replicas)`

Reply:
No worries, updated.